pkg/ruler/compat.go

```diff
@@ -263,14 +264,20 @@ func DefaultTenantManagerFactory(
 		wrappedQueryFunc = MetricsQueryFunc(queryFunc, totalQueries, failedQueries)
 		wrappedQueryFunc = RecordAndReportRuleQueryMetrics(wrappedQueryFunc, queryTime, logger)
 
+		externalURL := cfg.ExternalURL.URL
+		if externalURL == nil {
+			// We need a non-nil URL. dskit sets a nil URL when the yaml is `url: ""`.
+			// If we were to url.Parse(""), that will give a pointer to a zero-valued url.Url.
+			externalURL = &url.URL{}
+		}
 		return rules.NewManager(&rules.ManagerOptions{
 			Appendable:                 NewPusherAppendable(p, userID, overrides, totalWrites, failedWrites),
 			Queryable:                  embeddedQueryable,
 			QueryFunc:                  wrappedQueryFunc,
 			Context:                    user.InjectOrgID(ctx, userID),
 			GroupEvaluationContextFunc: FederatedGroupContextFunc,
-			ExternalURL:                cfg.ExternalURL.URL,
-			NotifyFunc:                 SendAlerts(notifier, cfg.ExternalURL.URL.String()),
+			ExternalURL:                externalURL,
```

Comment:
wouldn't it be safer to not rely on that `ExternalURL` is not `nil`?

Reply:
the code that uses this URL is just a single function:

https://github.com/grafana/mimir/blob/8188a22b1eff23f27edfe78727a7560c042897e6/vendor/github.com/prometheus/prometheus/template/template.go#L131-L140

I will submit a PR against prometheus to propose this

Reply:
interesting - the template doesn't cause a panic - it causes an error 

https://github.com/grafana/mimir/blob/8188a22b1eff23f27edfe78727a7560c042897e6/vendor/github.com/prometheus/prometheus/template/template.go#L374-L383

which will be included in the rendered template that the user sees 

https://github.com/grafana/mimir/blob/8fca3461c1e9f74cdbc98a8fc1fcfbbdcbbe24da/vendor/github.com/prometheus/prometheus/rules/alerting.go#L375


So I think prometheus handles nil well enough. In this case it should be enough for us to just prevent that single panic in the description of this PR. I pushed the change in 33755e459ff4cc248e3505507608643722ddda7b. What do you think?

Reply:
> What do you think?

Looks great, thanks for addressing my comment